File: out.py
#!/usr/bin/env python3

import json
import os
import sys
import boto3
import mypy_boto3_elasticbeanstalk as elasticbeanstalk


# Create the application version
def create_application_version(
    client: elasticbeanstalk.ElasticBeanstalkClient,
    application_name,
    version_label,
    s3_bucket,
    s3_key,
):
    client.create_application_version(
        ApplicationName=application_name,
        VersionLabel=version_label,
        SourceBundle={"S3Bucket": s3_bucket, "S3Key": s3_key},
    )


if __name__ == "__main__":
    stdin = sys.stdin.read()
    parsed = json.loads(stdin)
    # Print stdin to stderr
    print(f"stdin: {parsed}", file=sys.stderr)

    # Get the first command line argument and change the working directory to it
    os.chdir(sys.argv[1])

    # recurse through all directories in the current directory, printing them to stderr
    for root, dirs, files in os.walk("."):
        # Print the file as a full path
        for file in files:
            content = open(os.path.join(root, file)).read()
            print(f"{os.path.join(root, file)}: {content}", file=sys.stderr)
        # Print the directory as a full path
        for dir in dirs:
            print(os.path.join(root, dir), file=sys.stderr)
        
    client = boto3.client(
        "elasticbeanstalk",
        region_name=parsed["region"],
        aws_access_key_id=parsed["aws_access_key_id"],
        aws_secret_access_key=parsed["aws_secret_access_key"],
    )
    create_application_version(
        client,
        parsed["application_name"],
        parsed["version_label"],
        parsed["s3_bucket"],
        parsed["s3_key"],
    )

File: out.ts
#!/usr/bin/env ts-node-esm
import { wrapOut } from 'concourse-node-helper';
import { ElasticBeanstalk } from '@aws-sdk/client-elastic-beanstalk';
import { S3 } from '@aws-sdk/client-s3';

wrapOut<{
    aws_access_key_id: string,
    aws_secret_access_key: string,
    aws_region: string,
    application_name: string,
}, {
    version_label: string,
    bucket: string,
    key: string,
}>(async (input, ctx) => {
    // check for all the source params
    for (const param of ['aws_access_key_id', 'aws_secret_access_key', 'aws_region', 'application_name']) {
        if (!input.source[param]) {
            throw new Error(`Missing required source param: ${param}`);
        }
    }

    // Check for all the params
    for (const param of ['version_label', 'bucket', 'key']) {
        if (!input.params[param]) {
            throw new Error(`Missing required param: ${param}`);
        }
    }


    const ebClient = new ElasticBeanstalk({
        region: input.source.aws_region,
        credentials: {
            accessKeyId: input.source.aws_access_key_id,
            secretAccessKey: input.source.aws_secret_access_key
        }
    });

    try {

        const previouslyMade = await ebClient.describeApplicationVersions({
            ApplicationName: input.source.application_name,
            VersionLabels: [input.params.version_label]
        });

        if (previouslyMade.ApplicationVersions?.length ?? 0 > 0) {
            ctx.logit(`Version ${input.params.version_label} already exists, skipping`);
            return {
                version: {
                    ...input.version,
                    ref: previouslyMade.ApplicationVersions?.[0].VersionLabel ?? ''
                }
            }
        } else {
            const created = await ebClient.createApplicationVersion({
                ApplicationName: input.source.application_name,
                VersionLabel: input.params.version_label,
                SourceBundle: {
                    S3Bucket: input.params.bucket,
                    S3Key: input.params.key
                }
            });

            return {
                version: {
                    ...input.version,
                    ref: `${created.ApplicationVersion?.VersionLabel}`
                }
            }
        }
    } catch (e) {
        ctx.logit(`Error creating version: ${e.message}`);
        throw e;
    }

})()